packages/flutter/lib/src/material/bottom_navigation_bar.dart

```diff
@@ -0,0 +1,565 @@
+  /// The [BottomNavigationBar]'s [DestinationLabel]s have fixed width.
+  fixed,
+
+  /// The location and size of the [BottomNavigationBar] [DestinationLabel]s
+  /// animate larger when they are tapped.
+  shifting,
+}
+
+/// An interactive destination label within [BottomNavigationBar] with an icon
+/// and title.
+///
+/// See also:
+///
+///  * [BottomNavigationBar]
+///  * <https://material.google.com/components/bottom-navigation.html>
+class DestinationLabel {
+  /// Creates a label that is used with [BottomNavigationBar.labels].
+  ///
+  /// The arguments [icon] and [title] should not be null.
+  DestinationLabel({
+    @required this.icon,
+    @required this.title,
+    this.backgroundColor
+  }) {
+    assert(this.icon != null);
+    assert(this.title != null);
+  }
+
+  /// The icon of the label.
+  final Icon icon;
+
+  /// The title of the label.
+  final Widget title;
+
+  /// The color of the background radial animation.
+  ///
+  /// If the navigation bar's type is [BottomNavigationBarType.shifting], then
+  /// the entire bar is flooded with the [backgroundColor] when this label is
+  /// tapped.
+  final Color backgroundColor;
+}
+
+/// A material widget displayed at the bottom of an app for selecting among a
+/// small number of views.
+///
+/// The bottom navigation bar consists of multiple destinations in the form of
+/// labels laid out on top of a piece of material. It provies quick navigation
+/// between top-level views of an app and is typically used on mobile. For
+/// larger screens, side navigation may be a better fit.
+///
+/// A bottom navigation bar is usually used in conjunction with [Scaffold] where
+/// it is provided as the [Scaffold.bottomNavigationBar] argument.
+///
+/// See also:
+///
+///  * [DestinationLabel]
+///  * [Scaffold]
+///  * <https://material.google.com/components/bottom-navigation.html>
+class BottomNavigationBar extends StatefulWidget {
+  /// Creates a bottom navigation bar, typically used in a [Scaffold] where it
+  /// is provided as the [Scaffold.bottomNavigationBar] argument.
+  ///
+  /// The arguments [labels] and [type] should not be null.
+  ///
+  /// The number of labels passed should be equal or greater than 2.
+  ///
+  /// Passing a null [fixedColor] will cause a fallback to the theme's primary
+  /// color.
+  BottomNavigationBar({
+    Key key,
+    @required this.labels,
+    this.onTap,
+    this.currentIndex: 0,
+    this.type: BottomNavigationBarType.fixed,
+    this.fixedColor
+  }) : super(key: key) {
+    assert(this.labels != null);
+    assert(this.labels.length >= 2);
+    assert(0 <= currentIndex && currentIndex < this.labels.length);
+    assert(this.type != null);
+    assert(
+      this.type == BottomNavigationBarType.fixed || this.fixedColor == null
+    );
+  }
+
+  /// The interactive labels laid out within the bottom navigation bar.
+  final List<DestinationLabel> labels;
+
+  /// The callback that is called when a label is tapped.
```

Comment:
You need to explain that it's the responsibility of the callback's owner to keep track of the current index and to call setState() to rebuild its bottom bar.